### PR TITLE
Add DINOv3 vision foundation models contrib for Neuron

### PR DESCRIPTION
**Pull request by jimburtoft**
## Summary
- Add Meta DINOv3 self-supervised vision foundation models (ViT + ConvNeXt, 21M-6.7B params) to Neuron contrib
- ViT-S/B/L/H+ via `torch_neuronx.trace()`, ViT-7B (6.7B) via `neuronx-distributed` ModelBuilder TP=4
- ConvNeXt-T/B via `torch_neuronx.trace()` (exercises Conv2d, GroupNorm, depthwise conv)
- First encoder-only vision model with tensor parallelism on Neuron
- Validated on SDK 2.29 (torch-neuronx 2.9.0.2.13, neuronx-cc 2.24.5133.0)

## Key Results (SDK 2.29, trn2.3xlarge, LNC=2)

| Model | Params | 1-Core (img/s) | DP=4 Peak (img/s) | Cosine Sim |
|-------|--------|----------------:|-------------------:|------------|
| ViT-B/16 | 85.7M | 214.4 | **422.9** | 1.000000 |
| ConvNeXt-T | 27.8M | 264.4 | **522.6** | 0.999988 |
| ViT-7B/16 | 6.7B | OOM | **38.8** (TP=4) | Deterministic |

ConvNeXt-T improved 44% from SDK 2.28 to 2.29 (compiler improvements for conv ops).

## NKI Kernel Assessment (SDK 2.29 nki-lib)

Reviewed nki-lib 2.29 (23+ kernels). The library is LLM-inference-oriented:
- **LayerNorm**: `layernorm_tkg` kernel exists (direct fit for ViT) but compiler-generated version already performs well
- **Attention**: `attention_cte` designed for causal LLM prefill -- DINOv3 uses bidirectional (no causal mask). Identified gap.
- **MLP/SwiGLU**: NKI MLP kernels assume SwiGLU pattern; DINOv3 uses standard GELU FFN
- **Conv2d**: No NKI kernel for 2D convolution (ConvNeXt relies on compiler)

**Decision**: No NKI kernel integration. Compiler-generated kernels achieve strong performance (522 img/s ConvNeXt-T, 423 img/s ViT-B).

## Files
- `contrib/models/DINOv3/src/modeling_dinov3.py` -- Model loading, tracing, TP ViT-7B, accuracy validation, benchmarking (801 lines)
- `contrib/models/DINOv3/test/integration/test_model.py` -- 15 integration tests: smoke, accuracy, DataParallel, performance (366 lines)
- `contrib/models/DINOv3/README.md` -- Full documentation with results, GPU comparison, usage examples

## GPU Comparison (A10G g5.xlarge)
- ViT-B: Neuron 1.11x faster (422.9 vs 380.0 img/s)
- ConvNeXt-T: GPU 2.2x faster (1,156 vs 522.6 img/s) -- gap narrowed from 3.2x with SDK 2.29

## Testing
All 15 integration tests pass on trn2.3xlarge with SDK 2.29:
```
python -m pytest contrib/models/DINOv3/test/integration/test_model.py -v
```
Requires cloned DINOv3 repo: `git clone https://github.com/facebookresearch/dinov3.git /mnt/models/dinov3`

### CODE REVIEW

**Comment by jimburtoft** [comment]
Refreshed this on Inferentia2 with SDK 2.31 (neuronx-cc 2.26.6360, torch-neuronx 2.9.0.2.15), measured on inf2.xlarge and inf2.24xlarge. Rebased onto current main.

**Two fixes that were understating throughput**

`torch_neuronx.DataParallel` defaults `num_workers` to 2 regardless of core count, so `parallel_apply()`'s thread pool serializes all but two core dispatches. `benchmark_dataparallel()` now sets `num_workers = num_cores`:

| num_workers | ViT-L, 12 cores | Scaling |
|---:|---:|---:|
| 2 (default) | 155.7 img/s | 2.39x |
| 8 | 458.0 img/s | 7.04x |
| **12** | **806.2 img/s** | **12.38x** |

Also reverted an earlier `inline_weights=False` recommendation. That came from a trn2/SDK 2.30 async XLA error; on inf2 it costs 6.8x (ViT-L BS=1: 56.9 -> 8.4 img/s) because non-inlined weights are re-fed from the host every call.

**Maximum model size**

All seven variants run on inf2 and all of them fit on a single inf2.xlarge. ViT-7B (6.7B) needs TP>=2 -- at TP=1 the runtime reaches 15.947 GB of the 16 GB core and fails the next 64 MB allocation; at TP=2 it uses 14.72 GB (92%) and runs. TP must be a power of two (TP=12 is rejected: embed_dim 4096 not divisible by 12).

**Refreshed numbers** (inf2.xlarge, 2 cores, best of a workers x batch sweep)

ViT-S 832.4, ConvNeXt-T 501.9, ViT-B 421.4, ConvNeXt-B 275.7, ViT-L 130.8, ViT-H+ 69.1, ViT-7B 11.4 img/s. Accuracy cosine 1.000000 on all four ViTs, 0.99999 on ConvNeXt.

Batch size splits by model: small models peak at BS=1-4 and regress at BS=16, while large models must batch -- ViT-H+ gains 6.8x from BS=1 to BS=4. Per-core throughput on inf2.xlarge is 0.83-0.99x of inf2.24xlarge, so the 24xlarge no longer gets a separate table.

The older inf2 accuracy issue is also resolved upstream: DINOv3 now uses index slicing instead of `torch.split()` on dim>=2 in `Attention.apply_rope`, and cosine similarity is 1.000000 with no workaround.

**A negative result worth recording**

`src/nki_gelu_trace.py` is a custom NKI exact-erf GELU kernel, included but deliberately **not** wired in. GELU looked like the right target -- 34% of forward Vector+Scalar time, and neuronx-cc lowers exact erf as 1664 matmuls + 256 reciprocals where the kernel uses ~26 activation-table ops. The kernel is numerically exact but **slower in 13 of 14 model/batch configurations** (-5.8% to -14.5%), because the compiler already fuses GELU into the surrounding MLP GEMMs and a standalone kernel adds an HBM round-trip per call. Kept only as a working reference for the `nki_jit` convention under `torch_neuronx.trace()`, which differs from the standalone `nki` wheel in five ways documented in the file.

Worth flagging the methodology error behind it: an intermediate result claimed +32% for ViT-L at BS=1. That was a cold baseline, not a fast kernel -- measured back-to-back the ViT-L BS=1 baseline climbs 52.0 -> 62.4 -> 68.4 img/s while the NKI path is flat at 70.2 from the first run. Against a converged baseline the kernel is +2.8%, i.e. noise.

Also removes the A10G GPU comparison section, and adds `benchmark/` with the harnesses used above.

**Comment by jimburtoft** [comment]
Added the missing axis of the inference matrix: **PyTorch Native on Inferentia2**. Native had been measured on trn2 and `torch_neuronx.trace()` on inf2, but never Native on inf2.

**It works on gen2 and is accurate.** cos_sim 1.000000 eager/compiled FP32, 0.999996 compiled BF16, deterministic run-to-run. None of the documented trn1/gen2 Native gen-gates apply -- they are all training-path (grad-clip `sendrecv`, embedding-backward `hwdge`, multi-node CCOM topology).

**It is ~2x faster for ViT BF16 inference** (single core, best-of-3):

| Model | trace FP32+matmult | Native BF16 | Speedup |
|---|---:|---:|---:|
| ViT-S/16 | 416.2 | **949.2** | 2.28x |
| ViT-B/16 | 213.1 | **433.7** | 2.04x |
| ViT-L/16 | 71.8 | **136.2** | 1.90x |

Compiles are 5-10x faster too (ViT-L ~20s vs ~210s).

**The non-obvious part: this is a compilation-path effect that only appears in BF16.** I initially assumed it was mostly precision and that tracing BF16 weights would recover it. The 2x2 control says otherwise:

| | ViT-L | ViT-B |
|---|---:|---:|
| trace FP32 -> trace BF16 | 1.12x | 1.02x |
| trace BF16 -> Native BF16 (matched precision) | **1.70x** | **1.99x** |
| trace FP32 -> Native FP32 | **0.73x** | **0.86x** |

So tracing BF16 is not a substitute (only 2-12%, since `--auto-cast=matmult` already runs matmuls in bf16), and at FP32 Native is actually *slower* than trace.

**This contrib stays on `torch_neuronx.trace()` regardless**, for three reasons: ViT-7B's TP path uses `neuronx-distributed` ModelBuilder which is **not present on SDK 2.32**, and trace+NxD measured 1.83x *faster* than Native for ViT-7B TP elsewhere -- switching would regress the largest model and lose the max-model-size result; PyTorch Native is beta software in a private container whose `:latest` has repointed three times; and ConvNeXt/ViT-H+ were not tested on Native. The README documents Native as the faster option for ViT-only BF16 when beta tooling is acceptable.

One measurement note for anyone benchmarking this path: **PyTorch Native has asynchronous execution enabled by default.** Without `torch.neuron.synchronize()` in the timing loop, ViT-L BS=1 mismeasures at 1086 img/s / 0.92 ms -- a 15x overstatement, with a garbage cos_sim because the output is read before it exists. The benchmark script handles this and reports both synced latency and pipelined throughput.

**Comment by jimburtoft** [comment]
Re-measured the **Trainium2** numbers on SDK 2.31. The previous trn2 table was understated by **4.2-5.4x** -- it was measured with `DataParallel` at the broken `num_workers=2` default, at BS=1 only.

| Model | old DP=4 | LNC=2 (4 cores) | LNC=1 (8 cores) | vs old |
|---|---:|---:|---:|---:|
| ViT-S/16 | 722.8 | 3,780.4 | **3,893.7** | 5.39x |
| ViT-B/16 | 422.9 | 1,930.5 | **1,969.5** | 4.66x |
| ViT-L/16 | 174.7 | 348.8 | **739.7** | 4.23x |
| ConvNeXt-T | 522.6 | 2,310.9 | **2,401.8** | 4.60x |
| ConvNeXt-B | 257.8 | 1,285.6 | 1,298.6 | 5.04x |

**LNC=1 is the better default for every model tested.** The margin is small (1.00-1.04x) except **ViT-L, where LNC=1 is 2.12x faster** -- consistent with ViT-L being memory-bandwidth bound and benefiting from the smaller per-core NEFF. Same conclusion as the earlier LNC comparison, but reached without DataParallel in the loop.

The `num_workers` defect reproduces on trn2 (LNC=2, 4 cores, BS=1): ViT-S 651.4 -> 1216.0 (1.87x), ViT-B 400.1 -> 726.8 (1.82x), ViT-L 164.4 -> 305.8 (1.86x). Worth noting ViT-L's broken-default 164.4 closely reproduces the old table's 174.7, which confirms the historical numbers were measured with the defect rather than differing for some other reason.

**One trn2-specific trap documented while doing this:** `NEURON_CC_FLAGS="--lnc=1"` is **not** sufficient to set the LNC compile target here, because `trace_dinov3()` passes an explicit `compiler_args` list that overrides the env var. The NEFF silently compiles at the default `--lnc=2` and the runtime then rejects it at load:

```
NRT:nrt_load_util  Mismatch detected between Runtime configuration and NEFF.
    Runtime currently configured with `NEURON_LOGICAL_NC_CONFIG=1`
    but NEFF ... was compiled with `--lnc=2`.
```

The target has to go through `compiler_args` as `--logical-nc-config`. `benchmark/bench_trn2.py` handles this, and also precompiles in a single process before fanning out (N concurrent compiles contend for the runtime and fail init).

ViT-H+ and ViT-7B were not re-measured on trn2 this round; both are flagged as such in the README.

**Comment by jimburtoft** [comment]
Measured the multi-core PyTorch Native numbers on **inf2.24xlarge (12 NeuronCores)**. The previous Native evaluation was single-core with per-instance only *estimated* as 2x.

**Scaling is essentially perfect.** ViT-L BF16 BS=1, one worker process per core:

| Cores | img/s | Scaling | Efficiency | p50 (ms) |
|---:|---:|---:|---:|---:|
| 1 | 103.8 | 1.00x | 100.0% | 10.84 |
| 2 | 206.5 | 1.99x | 99.5% | 10.90 |
| 4 | 413.3 | 3.98x | 99.5% | 10.87 |
| 8 | 825.4 | 7.95x | 99.4% | 10.96 |
| **12** | **1,237.5** | **11.92x** | **99.3%** | 10.90 |

99.3% efficiency with latency flat at ~10.9 ms, so per-instance throughput is just (per-core rate) x (core count).

**The important correction: the verdict splits by architecture.** Per-instance on 12 cores, each side at its own best batch size:

| Model | trace | Native | Ratio | Use |
|---|---:|---:|---:|---|
| ViT-S/16 | 5,151.7 | **8,118.1** | 1.58x | Native |
| ViT-B/16 | 2,562.8 | **4,176.3** | 1.63x | Native |
| ViT-L/16 | 940.4 | **1,542.3** | 1.64x | Native |
| ConvNeXt-T | **3,192.2** | 2,550.9 | 0.80x | trace |
| ConvNeXt-B | **1,716.3** | 1,320.1 | 0.77x | trace |

**ConvNeXt is 1.25-1.30x FASTER on the traced path**, and batching does not rescue it on Native (ConvNeXt-T: 2,335.7 / 2,529.4 / 2,550.9 img/s at BS=1/4/8, still short of trace's 3,192.2). My earlier write-up flagged ConvNeXt as untested and implicitly assumed the ViT result would carry over -- it does not. The conv lowering evidently does not benefit from `torch.compile` the way the transformer ops do.

This strengthens the case for keeping trace as the contrib default: switching would regress ConvNeXt on top of losing the ViT-7B TP path (`neuronx-distributed` is absent on SDK 2.32).

**Comment by jimburtoft** [comment]
Closed the last two "not measured" items.

**trn2 SDK 2.31, LNC=2** (the two models still carrying SDK 2.28 figures):

| Model | Config | BS=1 | BS=4 |
|---|---|---:|---:|
| ViT-H+/16 | 4 independent procs | 11.2 | **119.8** |
| ViT-7B/16 | TP=4 | **34.9** | -- |

ViT-H+ gains **10.7x from BS=1 -> BS=4**. Its old 10.5 img/s figure was understated 11.4x, but the cause is worth stating precisely: BS=1 re-measures at **11.2**, almost exactly the old value, so the understatement came **entirely from not batching** -- not from the `num_workers` defect. ViT-H+ must be batched.

ViT-7B TP=4 reproduces at 34.9 vs the old 38.8 (-10%), so no meaningful change on 2.31. TP models dispatch from one process, so the `num_workers` defect never applied to them.

**ViT-H+ on PyTorch Native** (inf2.8xlarge, 2 cores) -- measured separately because ViT-H+ uses **SwiGLU** rather than `nn.GELU`, a different FFN path from ViT-S/B/L:

| Config | trace | Native | Ratio |
|---|---:|---:|---:|
| 1 core, BS=1 | 5.1 | **47.1** | **9.24x** |
| 1 core, BS=4 | 34.6 | **83.4** | 2.41x |
| 2 cores, best | 69.1 | **143.9** | **2.08x** |

**ViT-H+ is the strongest Native win in the suite -- 9.2x at batch 1.** The traced path lowers SwiGLU poorly at BS=1 and needs batching to recover; Native does not. It also scales at 86-92% efficiency across the 2 cores.

**The Native tradeoff is now fully mapped**: every ViT variant gains (1.58x-9.24x) while both ConvNeXt variants lose (0.77-0.80x). So the real choice is a mixed deployment rather than a clean switch -- and the contrib still can't move, because ViT-7B's TP path needs `neuronx-distributed`, which is absent on SDK 2.32.

Every model in the registry now has a current SDK 2.31 number on both inf2 and trn2, on both compilation paths where applicable.
